core/src/service/search_service.rs

```diff
@@ -49,4 +66,344 @@ impl RequestContext<'_, '_> {
+            should_continue.store(false, atomic::Ordering::Relaxed);
+
+            match search {
+                SearchRequest::Search { input } => {
+                    should_continue = Arc::new(AtomicBool::new(true));
+
+                    let results_tx = results_tx.clone();
+                    let files_info = files_info.clone();
+                    let should_continue = should_continue.clone();
+                    let input = input.clone();
+
+                    thread::spawn(move || {
+                        if let Err(search_err) =
+                            RequestContext::search(&results_tx, files_info, should_continue, input)
+                        {
+                            if let Err(err) = results_tx.send(SearchResult::Error(search_err)) {
+                                warn!("Send failed: {:#?}", err);
+                            }
+                        }
+                    });
+                }
+                SearchRequest::EndSearch => return Ok(()),
+                SearchRequest::StopCurrentSearch => {}
+            }
+        }
+    }
+
+    fn search(
+        results_tx: &Sender<SearchResult>, files_info: Arc<Vec<SearchableFileInfo>>,
+        should_continue: Arc<AtomicBool>, search: String,
+    ) -> Result<(), UnexpectedError> {
+        let mut no_matches = true;
+
+        RequestContext::search_file_names(
+            results_tx,
+            &should_continue,
+            &files_info,
+            &search,
+            &mut no_matches,
+        )?;
+
+        if should_continue.load(atomic::Ordering::Relaxed) {
+            RequestContext::search_file_contents(
+                results_tx,
+                &should_continue,
+                &files_info,
+                &search,
+                &mut no_matches,
+            )?;
+
+            if no_matches && should_continue.load(atomic::Ordering::Relaxed) {
+                results_tx.send(SearchResult::NoMatch)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn search_file_names(
+        results_tx: &Sender<SearchResult>, should_continue: &Arc<AtomicBool>,
+        files_info: &Arc<Vec<SearchableFileInfo>>, search: &str, no_matches: &mut bool,
+    ) -> Result<(), UnexpectedError> {
+        for info in files_info.as_ref() {
+            if !should_continue.load(atomic::Ordering::Relaxed) {
+                return Ok(());
+            }
+
+            if let Some(fuzzy_match) = FuzzySearch::new(search, &info.path)
+                .case_insensitive()
+                .score_with(&Scoring::emphasize_distance())
+                .best_match()
+            {
+                if fuzzy_match.score().is_positive() {
+                    if *no_matches {
+                        *no_matches = false;
+                    }
+
+                    if !should_continue.load(atomic::Ordering::Relaxed) {
+                        return Ok(());
+                    }
```

Comment:
Why do you do this check at the beginning of each iteration, and then here if the iteration is a match? Is at the beginning of the loop not enough?

Reply:
If `should_continue` becomes false after the start of the loop, I don't want it to send if it finds a match in that iteration. It will return an old query's result then.

Reply:
I see, ok.